File: papers/miniKanren/ocanren/unify/original.ml
open GT
open OCanren
open OCanren.Std
open Helper

let rec eq_nat a b q36 =
  fresh (q37) (q37 === (Pair.pair a b))
    (conde
       [(q37 === (Pair.pair (Nat.zero) (Nat.zero))) &&& (q36 === (!! true));
       fresh (q39) (q37 === (Pair.pair (Nat.succ q39) (Nat.zero))) (q36 === (!! false));
       fresh (q41) (q37 === (Pair.pair (Nat.zero) (Nat.succ q41))) (q36 === (!! false));
       fresh (x y) (q37 === (Pair.pair (Nat.succ x) (Nat.succ y))) (eq_nat x y q36)])

let rec get_term var subst q32 =
  ((subst === (nil ())) &&& (q32 === (none ()))) ||| (fresh (x xs) (subst === ((%) x xs)) (((var === (Nat.zero)) &&& (x === q32)) ||| (fresh (n) (var === (Nat.succ n)) (get_term n xs q32))))

let rec check_uni subst t1 (t2 : (ground_term, logic_term) Logic.injected) q31 =
  let rec forall2 subst l1 l2 q0 =
    fresh (q1) (q1 === (Pair.pair l1 l2))
      (((q1 === (Pair.pair (nil ()) (nil ()))) &&& (q0 === (!! true))) |||
         (fresh (x xs y ys q3 q4) (q1 === (Pair.pair ((%) x xs) ((%) y ys))) (
            check_uni subst x y q3) (forall2 subst xs ys q4) (conde [(q3 === (!! false)) &&& (q0 === (!! false)); (q3 === (!! true)) &&& (q0 === q4)]))) in
  fresh (q11) (q11 === (Pair.pair t1 t2))
    (conde
       [fresh (n1 a1 n2 a2 q12 q13) (q11 === (Pair.pair (constr n1 a1) (constr n2 a2))) (
          eq_nat n1 n2 q12) (forall2 subst a1 a2 q13) (conde [(q12 === (!! false)) &&& (q31 === (!! false)); (q12 === (!! true)) &&& (q31 === q13)]);
       fresh (v n a q19) (q11 === (Pair.pair (var_ v) (constr n a))) (get_term v subst q19)
         (((q19 === (none ())) &&& (q31 === (!! false))) ||| (fresh (t) (q19 === (some t)) (check_uni subst t t2 q31)));
       fresh (n a v q22) (q11 === (Pair.pair (constr n a) (var_ v))) (get_term v subst q22)
         (((q22 === (none ())) &&& (q31 === (!! false))) ||| (fresh (t) (q22 === (some t)) (check_uni subst t1 t q31)));
       fresh (v1 v2 q25) (q11 === (Pair.pair (var_ v1) (var_ v2))) (get_term v1 subst q25)
         ((fresh (t1') (q25 === (some t1')) (check_uni subst t1' t2 q31)) |||
            (fresh (q27) (q25 === (none ())) (get_term v2 subst q27) ((fresh (q28) (q27 === (some q28)) (q31 === (!! false))) ||| ((q27 === (none ())) &&& (eq_nat v1 v2 q31)))))])

File: papers/aplas/benchmarks/unify/original.pl
eq_nat(z, z, true).
eq_nat(s(Q39), z, false).
eq_nat(z, s(Q41), false).
eq_nat(s(X), s(Y), Q36) :- eq_nat(X, Y, Q36).
get_term(Var, nil, none).
get_term(z, cons(Q32, Xs), Q32).
get_term(s(N), cons(X, Xs), Q32) :- get_term(N, Xs, Q32).
check_uni(Subst, constr(N1, A1), constr(N2, A2), false) :- eq_nat(N1, N2, false), forall2(Subst, A1, A2, Q13).
check_uni(Subst, constr(N1, A1), constr(N2, A2), Q13) :- eq_nat(N1, N2, true), forall2(Subst, A1, A2, Q13).
check_uni(Subst, var_(V), constr(N, A), false) :- get_term(V, Subst, none).
check_uni(Subst, var_(V), constr(N, A), Q31) :- get_term(V, Subst, some(T)), check_uni(Subst, T, constr(N, A), Q31).
check_uni(Subst, constr(N, A), var_(V), false) :- get_term(V, Subst, none).
check_uni(Subst, constr(N, A), var_(V), Q31) :- get_term(V, Subst, some(T)), check_uni(Subst, constr(N, A), T, Q31).
check_uni(Subst, var_(V1), var_(V2), Q31) :- get_term(V1, Subst, some(T11)), check_uni(Subst, T11, var_(V2), Q31).
check_uni(Subst, var_(V1), var_(V2), false) :- get_term(V1, Subst, none), get_term(V2, Subst, some(Q28)).
check_uni(Subst, var_(V1), var_(V2), Q31) :- get_term(V1, Subst, none), get_term(V2, Subst, none), eq_nat(V1, V2, Q31).
get_term(Var, nil, none).
get_term(z, cons(Q32, Xs), Q32).
get_term(s(N), cons(X, Xs), Q32) :- get_term(N, Xs, Q32).
eq_nat(z, z, true).
eq_nat(s(Q39), z, false).
eq_nat(z, s(Q41), false).
eq_nat(s(X), s(Y), Q36) :- eq_nat(X, Y, Q36).
forall2(Subst, nil, nil, true).
forall2(Subst, cons(X, Xs), cons(Y, Ys), false) :- check_uni(Subst, X, Y, false), forall2(Subst, Xs, Ys, Q4).
forall2(Subst, cons(X, Xs), cons(Y, Ys), Q4) :- check_uni(Subst, X, Y, true), forall2(Subst, Xs, Ys, Q4).
forall2(Subst, nil, nil, true).
forall2(Subst, cons(X, Xs), cons(Y, Ys), false) :- check_uni(Subst, X, Y, false), forall2(Subst, Xs, Ys, Q4).
forall2(Subst, cons(X, Xs), cons(Y, Ys), Q4) :- check_uni(Subst, X, Y, true), forall2(Subst, Xs, Ys, Q4).
check_uni(Subst, constr(N1, A1), constr(N2, A2), false) :- eq_nat(N1, N2, false), forall2(Subst, A1, A2, Q13).
check_uni(Subst, constr(N1, A1), constr(N2, A2), Q13) :- eq_nat(N1, N2, true), forall2(Subst, A1, A2, Q13).
check_uni(Subst, var_(V), constr(N, A), false) :- get_term(V, Subst, none).
check_uni(Subst, var_(V), constr(N, A), Q31) :- get_term(V, Subst, some(T)), check_uni(Subst, T, constr(N, A), Q31).
check_uni(Subst, constr(N, A), var_(V), false) :- get_term(V, Subst, none).
check_uni(Subst, constr(N, A), var_(V), Q31) :- get_term(V, Subst, some(T)), check_uni(Subst, constr(N, A), T, Q31).
check_uni(Subst, var_(V1), var_(V2), Q31) :- get_term(V1, Subst, some(T11)), check_uni(Subst, T11, var_(V2), Q31).
check_uni(Subst, var_(V1), var_(V2), false) :- get_term(V1, Subst, none), get_term(V2, Subst, some(Q28)).
check_uni(Subst, var_(V1), var_(V2), Q31) :- get_term(V1, Subst, none), get_term(V2, Subst, none), eq_nat(V1, V2, Q31).
get_term(Var, nil, none).
get_term(z, cons(Q32, Xs), Q32).
get_term(s(N), cons(X, Xs), Q32) :- get_term(N, Xs, Q32).
eq_nat(z, z, true).
eq_nat(s(Q39), z, false).
eq_nat(z, s(Q41), false).
eq_nat(s(X), s(Y), Q36) :- eq_nat(X, Y, Q36).
forall2(Subst, nil, nil, true).
forall2(Subst, cons(X, Xs), cons(Y, Ys), false) :- check_uni(Subst, X, Y, false), forall2(Subst, Xs, Ys, Q4).
forall2(Subst, cons(X, Xs), cons(Y, Ys), Q4) :- check_uni(Subst, X, Y, true), forall2(Subst, Xs, Ys, Q4).